### PR TITLE
Getting 401 when Community or Collection Administrator tries to edit Collection Authorization Group

### PR DESCRIPTION
**Pull request by Raf-atmire**
## References
* Fixes https://github.com/DSpace/dspace-angular/issues/927

## Description
This PR fixes the issues listed in the above mentioned issue, as described by Art here: https://github.com/DSpace/dspace-angular/issues/927#issuecomment-777505957

## Instructions for Reviewers

List of changes in this PR:
* Moved the access-control out of the admin module
* Created guards for the edit community and collections
* Used the guards on the routes
* Showing and hiding of the sidebar edits based on the features implemented in the accompanying REST PR: https://github.com/DSpace/DSpace/pull/3148

How to test this PR:
* Setup this branch locally with the backend being the branch listed here: https://github.com/DSpace/DSpace/pull/3148
* Create an EPerson that is an admin of a specific community or collection
* Run index discovery
* Login as admin, verify that you can view community and collection edits in the sidebar as well as the edit buttons on the specific community and collection pages. Verify that actually editting them works as well
* Login as the EPerson that is an admin of the specific community or collection. Verify that the sidebar edit sections for communities and collections are present. Verify that there is an edit button on the community/collection page for which the EPerson is an admin, but not on other community/collection pages. Verify that editting works on the community/collection that the eperson is an admin
* Login as a regular user, verify that the edit community and collection sidebar sections aren't present, verify that the edit buttons on community/collection pages aren't present. Verify that manually traversing to an edit url gives a 403

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [x] My PR doesn't introduce circular dependencies
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.

### CODE REVIEW

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Raf-atmire : A recent change on `main` seems to have broken the ability to build this PR.  Now, if I try to build this using `yarn start` or `yarn start:dev`, I get this error...
```
ERROR in src/themes/custom/theme.module.ts:3:42 - error TS2307: Cannot find module '../../app/+admin/admin-access-control/admin-access-control.module' or its corresponding type declarations.

3 import { AdminAccessControlModule } from '../../app/+admin/admin-access-control/admin-access-control.module';
```

Could you rebase this on main & fix the issue?  It appears like it might be some sort of conflict with the theming work (in #1035) which was just merged, even though it's not showing up as a code conflict in GitHub.  I'll verified this same issue doesn't occur when building current `main`. It only occurs when building this PR applied on top of `main`.

**Comment by Raf-atmire** [comment]
@tdonohue Thank you for the feedback! I've merged the dspace/main into the branch, could you please check again? Thanks!

**Comment by tdonohue** [comment]
@Raf-atmire : It looks like this PR has build issues now....here's the error as reported by our GitHub Actions CI:
```
ERROR in src/themes/custom/theme.module.ts:3:42 - error TS2307: Cannot find module '../../app/+admin/admin-access-control/admin-access-control.module' or its corresponding type declarations.

3 import { AdminAccessControlModule } from '../../app/+admin/admin-access-control/admin-access-control.module';
```

File: src/app/+admin/admin-sidebar/admin-sidebar.component.ts [hunk 1]
```diff
@@ -102,192 +99,209 @@ export class AdminSidebarComponent extends MenuComponent implements OnInit {
...
+            text: 'menu.section.new'
+          } as TextMenuItemModel,
+          icon: 'plus-circle',
+          index: 0
+        },
+        {
+          id: 'new_community',
+          parentID: 'new',
+          active: false,
+          visible: true,
```
Comment (tdonohue) [inline src/app/+admin/admin-sidebar/admin-sidebar.component.ts:135]: It seems like both `new_community` and `new_collection` (just below) should be limited to `IsCommunityAdmin` as well.  You cannot create either of those objects unless you are a CommunityAdmin or full Admin.

File: src/app/+admin/admin-sidebar/admin-sidebar.component.ts [hunk 2]
```diff
@@ -102,192 +99,209 @@ export class AdminSidebarComponent extends MenuComponent implements OnInit {
...
+            function: () => {
+              this.modalService.open(CreateItemParentSelectorComponent);
+            }
+          } as OnClickMenuItemModel,
+        },
+        {
+          id: 'new_process',
+          parentID: 'new',
+          active: false,
+          visible: true,
```
Comment (tdonohue) [inline src/app/+admin/admin-sidebar/admin-sidebar.component.ts:174]: Processes can only be run by someone who is a CollectionAdmin, CommunityAdmin or full admin.  Should we change this to check if you are (minimally) a CollectionAdmin?

File: src/app/+admin/admin-sidebar/admin-sidebar.component.ts [hunk 3]
```diff
@@ -102,192 +99,209 @@ export class AdminSidebarComponent extends MenuComponent implements OnInit {
...
+            function: () => {
+              this.modalService.open(CreateCollectionParentSelectorComponent);
+            }
+          } as OnClickMenuItemModel,
+        },
+        {
+          id: 'new_item',
+          parentID: 'new',
+          active: false,
+          visible: true,
```
Comment (tdonohue) [inline src/app/+admin/admin-sidebar/admin-sidebar.component.ts:161]: It seems like `new_item` should be limited to people who have submit access. Is there a check for that we could add in here? (If not, maybe this needs to be a separate ticket)

File: src/app/+admin/admin-sidebar/admin-sidebar.component.ts [hunk 4]
```diff
@@ -102,192 +99,209 @@ export class AdminSidebarComponent extends MenuComponent implements OnInit {
...
+            function: () => {
+              this.modalService.open(EditCollectionSelectorComponent);
+            }
+          } as OnClickMenuItemModel,
+        },
+        {
+          id: 'edit_item',
+          parentID: 'edit',
+          active: false,
+          visible: true,
```
Comment (tdonohue) [inline src/app/+admin/admin-sidebar/admin-sidebar.component.ts:235]: We might want to limit this to isCollectionAdmin too, just for now (until we have more fine grained Item edit permissions)

File: src/app/+admin/admin-sidebar/admin-sidebar.component.ts [hunk 5]
```diff
@@ -102,192 +99,209 @@ export class AdminSidebarComponent extends MenuComponent implements OnInit {
...
-          link: ''
-        } as LinkMenuItemModel,
-        icon: 'chart-bar',
-        index: 8
-      },
+        /* Curation tasks */
+        {
+          id: 'curation_tasks',
+          active: false,
+          visible: true,
```
Comment (tdonohue) [inline src/app/+admin/admin-sidebar/admin-sidebar.component.ts:249]: Curation Tasks also require CollectionAdmin, CommunityAdmin or full Admin

File: src/app/+admin/admin-sidebar/admin-sidebar.component.ts [hunk 6]
```diff
@@ -102,192 +99,209 @@ export class AdminSidebarComponent extends MenuComponent implements OnInit {
...
-      },
-    ];
-    menuList.forEach((menuSection) => this.menuService.addSection(this.menuID, Object.assign(menuSection, {
-      shouldPersistOnRouteChange: true
-    })));
+        /* Control Panel */
+        {
+          id: 'control_panel',
+          active: false,
+          visible: true,
```
Comment (tdonohue) [inline src/app/+admin/admin-sidebar/admin-sidebar.component.ts:277]: This should be full Admin only (even though it doesn't yet exist)

File: src/app/+admin/admin-sidebar/admin-sidebar.component.ts [hunk 7]
```diff
@@ -102,192 +99,209 @@ export class AdminSidebarComponent extends MenuComponent implements OnInit {
...
+          } as LinkMenuItemModel,
+          icon: 'cogs',
+          index: 9
+        },
+
+        /* Processes */
+        {
+          id: 'processes',
+          active: false,
+          visible: true,
```
Comment (tdonohue) [inline src/app/+admin/admin-sidebar/admin-sidebar.component.ts:291]: Processes should be Admin only as well (I believe).  Definitely not enabled for everyone.

**Review by tdonohue — CHANGES_REQUESTED** [review]
@Raf-atmire : Overall, this looks good & it works (I tested it with the corresponding REST API PR).
I did notice though that when logging in as a user with *zero permissions*, I see a lot of menu options that are invalid for me.  So, I made some suggestions inline below on how we might access restrict additional menu options.  Some of these changes _might be_ temporary, until we find a need for more fine-grained permission checks for some menu items.  But, I'd rather they be restricted *now* instead of appearing publicly.

Otherwise, honestly this all looks good to me.  It just needs those minor changes (or feel free to suggest alternatives if you feel these changes don't belong here) & the corresponding REST API PR still has feedback needing to be addressed.

**Comment by tdonohue** [comment]
@ybnd or @Raf-atmire : Just a quick note that this PR now seems to have a few small merge conflicts (likely from something merged late last week).  Once those are resolved, I'll gladly re-test. Thanks!

File: src/app/+admin/admin-sidebar/admin-sidebar.component.ts (hunk 3 above)
Comment (Raf-atmire) [reply src/app/+admin/admin-sidebar/admin-sidebar.component.ts:162]: @ybnd and me couldn't find a feature that suffices for this, so we should probably provide another ticket for this

**Review by tdonohue — APPROVED** [review]
👍 Looks great! Thanks @Raf-atmire and @ybnd !  I tested this with the REST API backend and it's working exactly as described.

**Pull request merged**